src/platforms/linux/xdgstartatbootwatcher.cpp

```diff
@@ -0,0 +1,156 @@
+  m_token = "mozillavpn_" + QString::number(randbits, 16);
+  m_replyPath = xdgReplyPath();
+  QDBusConnection::sessionBus().connect(XDG_PORTAL_SERVICE, m_replyPath,
+                                        XDG_PORTAL_REQUEST, "Response", this,
+                                        SLOT(xdgResponse(uint, QVariantMap)));
+
+  startAtBootChanged();
+}
+
+XdgStartAtBootWatcher::~XdgStartAtBootWatcher() {
+  MZ_COUNT_DTOR(XdgStartAtBootWatcher);
+}
+
+// See the org.freedesktop.portal.Request documentation, the request portal will
+// likely take the form "/org/freedesktop/portal/desktop/request/<SENDER>/TOKEN"
+// and we should connect to it before trying to make a request.
+QString XdgStartAtBootWatcher::xdgReplyPath() {
+  QDBusConnection bus = QDBusConnection::sessionBus();
+  QString sender = bus.baseService().mid(1).replace('.', '_');
+  return QString("/org/freedesktop/portal/desktop/request/%1/%2")
+      .arg(sender)
+      .arg(m_token);
+}
+
+void XdgStartAtBootWatcher::xdgResponse(uint response, QVariantMap results) {
+  logger.debug() << "StartAtBoot responded:" << response;
+  for (auto i = results.cbegin(); i != results.cend(); i++) {
+    logger.debug() << "StartAtBoot" << QString("%1:").arg(i.key())
+                   << i.value().toString();
+  }
+}
+
+// Try to find the window identifier for an XDG desktop portal request.
+// https://flatpak.github.io/xdg-desktop-portal/docs/window-identifiers.html
+QString XdgStartAtBootWatcher::parentWindow() {
+  if (!QmlEngineHolder::exists()) {
+    return QString("");
+  }
+  QmlEngineHolder* holder = QmlEngineHolder::instance();
+  if (!holder->hasWindow()) {
+    return QString("");
+  }
+
+#if QT_VERSION >= QT_VERSION_CHECK(6, 5, 0)
+  QGenericUnixServices* services = dynamic_cast<QGenericUnixServices*>(
+      QGuiApplicationPrivate::platformIntegration()->services());
+  if (services != nullptr) {
+    return services->portalWindowIdentifier(holder->window());
+  }
+#else
+  // X11 is the only platform that we can get a window handle on prior to 6.5.0
+  if (QGuiApplication::platformName() == "xcb") {
+    return "x11:" + QString::number(holder->window()->winId(), 16);
+  }
+#endif
+
+  // Otherwise, we don't support this windowing system.
+  return QString("");
+}
+
+void XdgStartAtBootWatcher::startAtBootChanged() {
+  bool startAtBoot = SettingsHolder::instance()->startAtBoot();
+
+  logger.debug() << "StartAtBoot changed:" << startAtBoot;
+
+  QStringList cmdline = {QCoreApplication::applicationFilePath(), "ui", "-m",
+                         "-s"};
+  QVariantMap options;
+  options["autostart"] = QVariant(startAtBoot);
+  options["background"] = QVariant(true);
+  options["commandline"] = QVariant(cmdline);
+  options["handle_token"] = QVariant(m_token);
+
+  QDBusMessage call = QDBusMessage::createMethodCall(
+      XDG_PORTAL_SERVICE, XDG_PORTAL_PATH, XDG_PORTAL_BACKGROUND,
+      "RequestBackground");
+  call << parentWindow();
```

Comment:
Seems this can be "" if the windowing system is unsupported. Does it still make sense to do the call, or should this be checked and aborted if not?

Reply:
The XDG specs go out of their way to say that the window handle should be an empty string if the windowing system is unsupported. The purpose of the window handle in this case is to tell the windowing system where to display a UAC-style popup message if the action requires user approval, and if we provide an empty string, the window system just throws it in the middle of the screen.

Note that I have yet to find any desktop environment that actually displays such a prompt for this permission, so this code is all just theoretical anyways.